tests/tests.rs

```diff
@@ -451,3 +451,25 @@ fn par_join_many_entities_and_systems() {
                n);
     }
 }
+
+fn par_join_check_storage() {
+    use rayon::iter::ParallelIterator;
+
+    struct Sys;
+
+    let mut world = create_world();
+
+    impl<'a> System<'a> for Sys {
+        type SystemData = (WriteStorage<'a, CompInt>);
+
+        fn run(&mut self, data: Self::SystemData) {
+            let mut ints = data;
+            (ints.check())
+                .par_join()
+                .for_each(|c_i| ints.get_mut_unchecked(&mut c_i).0 += 1);
```

Comment:
If you switch `ints.get_mut_unchecked(&mut c_i).0 += 1` to `ints.get_unchecked(&c_i);` it works fine.
If you change it to `&mut ints;` then it errors: 'cannot borrow data mutably in a captured outer variable in an `Fn` closure', because `for_each` takes `Fn` closure as parameter.
If you assign `&mut ints` to variable and try to capture it, it will 'cannot borrow `ints` as immutable because it is also borrowed as mutable' (as it should).

I think the `get_mut_unchecked` should be able to take `&self` as parameter as `&mut Entry` and it's lifetime should ensure correctness.

This would force to wrap `data` field inside `Storage` to `UnsafeCell` (or `inner` inside `MaskedStorage`)...